SingularityService/src/main/java/com/hubspot/singularity/scheduler/SingularityUpstreamChecker.java

```diff
@@ -0,0 +1,170 @@
+      return new ArrayList<>(activeTasksForRequest.values());
+    }
+    LOG.error("TaskId not found for requestId: {}.", requestId);
+    throw new Exception("TaskId not found.");
+  }
+
+  private Collection<UpstreamInfo> getUpstreamsFromActiveTasksForRequest(String singularityRequestId, Optional<String> loadBalancerUpstreamGroup) throws Exception {
+    return lbClient.getUpstreamsForTasks(getActiveHealthyTasksForRequest(singularityRequestId), singularityRequestId, loadBalancerUpstreamGroup);
+  }
+
+  private boolean isEqualUpstreamGroupRackId(UpstreamInfo upstream1, UpstreamInfo upstream2){
+    return (upstream1.getUpstream().equals(upstream2.getUpstream()))
+        && (upstream1.getGroup().equals(upstream2.getGroup()))
+        && (upstream1.getRackId().equals(upstream2.getRackId()));
+  }
+
+  /**
+   * @param upstream
+   * @param upstreams
+   * @return a collection of upstreams in the upstreams param that match with the upstream param on upstream, group and rackId
+   * We expect that the collection will have a maximum of one match, but we will keep it as a collection just in case
+   */
+  private Collection<UpstreamInfo> getEqualUpstreams(UpstreamInfo upstream, Collection<UpstreamInfo> upstreams) {
+    return upstreams.stream().filter(candidate -> isEqualUpstreamGroupRackId(candidate, upstream)).collect(Collectors.toList());
+  }
+
+  private List<UpstreamInfo> getExtraUpstreams(Collection<UpstreamInfo> upstreamsInBaragonForRequest, Collection<UpstreamInfo> upstreamsInSingularityForRequest) {
+    for (UpstreamInfo upstreamInSingularity : upstreamsInSingularityForRequest) {
+      final Collection<UpstreamInfo> matches = getEqualUpstreams(upstreamInSingularity, upstreamsInBaragonForRequest);
+      upstreamsInBaragonForRequest.removeAll(matches);
+    }
+    return new ArrayList<>(upstreamsInBaragonForRequest);
+  }
+
+  private SingularityLoadBalancerUpdate syncUpstreamsForService(SingularityRequest singularityRequest, SingularityDeploy deploy, Optional<String> loadBalancerUpstreamGroup) throws Exception {
+    final String singularityRequestId = singularityRequest.getId();
+    final LoadBalancerRequestId loadBalancerRequestId = new LoadBalancerRequestId(singularityRequestId, LoadBalancerRequestType.REMOVE, Optional.absent());
+    Collection<UpstreamInfo> upstreamsInBaragonForRequest = lbClient.getLoadBalancerUpstreamsForRequest(loadBalancerRequestId.toString());
+    Collection<UpstreamInfo> upstreamsInSingularityForRequest = getUpstreamsFromActiveTasksForRequest(singularityRequestId, loadBalancerUpstreamGroup);
+    final List<UpstreamInfo> extraUpstreams = getExtraUpstreams(upstreamsInBaragonForRequest, upstreamsInSingularityForRequest);
+    LOG.info("Making and sending load balancer request to remove {} extra upstreams.", extraUpstreams.size());
+    LOG.info("Upstreams removed are: {}", Arrays.toString(extraUpstreams.toArray()));
```

Comment:
the logger will format this for you, can just put `extraUpstreams` as the arg here